prusti-viper/src/encoder/snapshot/encoder.rs

```diff
@@ -586,6 +630,162 @@ impl SnapshotEncoder {
                 self.encode_complex(encoder, variants, predicate_name)
             }
 
+            ty::TyKind::Array(elem_ty, ..) => {
+                let elem_snap_ty = self.encode_type(encoder, elem_ty)?;
+                let at = encoder.encode_array_types(ty)?;
+
+                let domain_name = format!("Snap${}", &at.array_pred);
+                let snap_type = Type::Snapshot(at.array_pred.clone());
+                let seq_type = Type::Seq(box elem_snap_ty.clone());
+
+                let cons = vir::DomainFunc {
+                    name: format!("cons${}$", domain_name),
+                    formal_args: vec![
+                        vir_local!{ data: {seq_type.clone()} },
+                    ],
+                    return_type: snap_type.clone(),
+                    unique: false,
+                    domain_name: domain_name.clone(),
+                };
+
+                let snap_body = cons.apply(
+                    vec![
+                        Expr::Seq(
+                            Type::Seq(box elem_snap_ty.clone()),
+                            (0..at.array_len)
+                                .into_iter()
+                                .map(|idx| {
+                                    at.encode_lookup_pure_call(
+                                        encoder,
+                                        arg_expr.clone(),
+                                        idx.into(),
+                                        elem_snap_ty.clone(),
+                                    )
+                                })
+                                .collect(),
+                            vir::Position::default(),
+                        )
+                    ]
+                );
+
+                let snap_func = vir::Function {
+                    name: SNAP_FUNC_NAME.to_string(),
+                    formal_args: vec![arg_self],
+                    return_type: snap_type.clone(),
+                    pres: vec![
+                        Expr::predicate_access_predicate(
+                            at.array_pred,
+                            arg_expr,
+                            PermAmount::Read,
+                        ),
+                    ],
+                    posts: vec![],
+                    body: Some(snap_body),
+                };
+
+                let constructor_inj = {
+                    let lhs_arg = vir_local!{ _l_data: {seq_type.clone()} };
+                    let rhs_arg = vir_local!{ _r_data: {seq_type.clone()} };
+
+                    let lhs_call = Expr::domain_func_app(
+                        cons.clone(),
+                        vec![lhs_arg.clone().into()],
+                    );
+                    let rhs_call = Expr::domain_func_app(
+                        cons.clone(),
+                        vec![rhs_arg.clone().into()],
+                    );
```

Comment:
IIRC I also added `.apply` to domain functions?

Reply:
ah, yes.. i even used it (two reviews above) :laughing: